autoPyTorch/api/results_manager.py

```diff
@@ -0,0 +1,318 @@
+
+        Args:
+            run_history (RunHistory):
+                The history of config evals from SMAC.
+        """
+
+        self.clear()  # Delete cache before the extraction
+
+        for run_key, run_value in run_history.data.items():
+            config_id = run_key.config_id
+            config = run_history.ids_config[config_id]
+
+            status_msg = STATUS2MSG.get(run_value.status, None)
+            if run_value.status in (StatusType.STOP, StatusType.RUNNING):
+                continue
+            elif status_msg is None:
+                raise ValueError(f'Unexpected run status: {run_value.status}')
+
+            is_traditional = False  # If run is not successful, unsure ==> not True ==> False
+            if run_value.additional_info is not None:
+                is_traditional = run_value.additional_info['configuration_origin'] == 'traditional'
+
+            self.update(
+                status=status_msg,
+                config=config,
+                budget=run_key.budget,
+                fit_time=run_value.time,  # To ravin: I added this line since fit_times was untouched.
+                score=cost2metric(cost=run_value.cost, metric=self._metric),
+                metric_info=_extract_metrics_info(run_value=run_value, scoring_functions=self._scoring_functions),
+                is_traditional=is_traditional,
+                additional_info=run_value.additional_info,
+                config_id=config_id
+            )
+
+        self.rank_test_scores = scipy.stats.rankdata(
+            -1 * self._metric._sign * self.opt_scores,  # rank order
+            method='min'
+        )
+
+
+class ResultsManager:
+    def __init__(self, *args: Any, **kwargs: Any):
+        """
+        Attributes:
+            run_history (RunHistory):
+                A `SMAC Runshistory <https://automl.github.io/SMAC3/master/apidoc/smac.runhistory.runhistory.html>`_
+                object that holds information about the runs of the target algorithm made during search
+            ensemble_performance_history (List[Dict[str, Any]]):
+                The list of ensemble performance in the optimization.
+                The list includes the `timestamp`, `result on train set`, and `result on test set`
+            trajectory (List[TrajEntry]):
+                A list of all incumbent configurations during search
+        """
+        self.run_history: RunHistory = RunHistory()
+        self.ensemble_performance_history: List[Dict[str, Any]] = []
+        self.trajectory: List[TrajEntry] = []
+
+    def _check_run_history(self) -> None:
+        if self.run_history is None:
+            raise RuntimeError("No Run History found, search has not been called.")
+
+        if self.run_history.empty():
+            raise RuntimeError("Run History is empty. Something went wrong, "
+                               "SMAC was not able to fit any model?")
+
+    def get_incumbent_results(
+        self,
+        metric: autoPyTorchMetric,
```

Comment:
can we have this metric argument as`Optional` ? I think the default behaviour of this function should be to return the incumbent result according to the `self._metric`.